lib/src/commit_builder.rs

```diff
@@ -61,7 +62,11 @@ impl CommitBuilder<'_> {
         }
     }
 
-    pub fn for_rewrite_from<'repo>(
+    // Use MutRepo::rewrite_commit() instead
+    //
+    // If we want to make this public, it should take a callback for
+    // `CommitBuilder::write` to call.
+    pub(crate) fn for_rewrite_from<'repo>(
```

Comment:
Out of curiosity, what kind of callback would be passed?

`CommitBuilder` takes mutable reference, so the `MutableRepo` is controlled solely by the `CommitBuilder`. I think it justifies the `.write()` API.

FWIW, making these functions `pub(crate)` makes sense.

Reply:
This is very helpful, thanks. I didn't think about the pointer ownership, and thinking about it makes me happier about the organization of code between these types.

I think you're also suggesting that the callback I was talking about would be difficult or impossible to write, which sounds likely to me.

I'll delete this paragraph from the comment.

Reply:
> the callback I was talking about would be difficult or impossible to write

The callback could take `&mut mut_repo` as a parameter if that makes sense.

FWIW, if we want to hide some MutableRepo functions (notably `record_*_commit()`), maybe we can add a `Wrapper(&mut MutableRepo)` that can be constructed only in the repo module, and pass it in place of `mut_repo`.

Reply:
> FWIW, if we want to hide some MutableRepo functions (notably record_*_commit()), maybe we can add a Wrapper(&mut MutableRepo) that can be constructed only in the repo module, and pass it in place of mut_repo.

I didn't quite understand what you meant. What would the wrapper do?

Currently, we can't really hide `record_*_commit`, the `split` command really needs pretty intimate access to them. However, as I said in the other comment, I'm not sure what the end state will look like exactly yet, but a wrapper is definitely an option.

What I do know I want at the moment:

- MutRepo should keep track of the `parent_mapping` of all the rebases that previously happened there (including the ones done by DescendantRebaser).
- There should be some way to request a rebase of a *subset* of commits as though some other commits were rewritten. One option I'm considering is to reuse `RebaseOptions` for this.

Reply:
> What would the wrapper do?

Suppose `write_commit()` and `record_rewritten_commit()` should be called only by `CommitBuilder`, these `pub` methods could be moved to `Handle(&mut MutableRepo)` type, and `MutableRepo` controls who can obtain the `Handle` type.

---

I have no idea about the `parent_mapping` stuff (simply because I don't follow the recent discussion about rebasing API problems.)

Reply:
> maybe we can add a Wrapper(&mut MutableRepo) that can be constructed only in the repo module, and pass it in place of mut_repo.

This was, in fact, what I was originally thinking. On third thought, maybe it won't violate ownership rules. I hadn't thought it through.

> Suppose write_commit() and record_rewritten_commit() should be called only by CommitBuilder...

Things like `record_rewritten_commit` need to be called by `jj split`, and likely more commands in the future.

I'll keep the idea of a Handle with a private constructor and public methods in mind.

Reply:
> Things like `record_rewritten_commit` need to be called by `jj split`, and likely more commands in the future.

I don't think CLI should depend on these low-level functions. Perhaps, the `jj split` case could be handled by `CommitBuilder` (by somehow teaching that the one of the split commit won't be the rebase destination.) For the other future improvements, maybe we'll need to extract rebase machinery to the lib? Ideally, the CLI part should only deal with user inputs and outputs.

Reply:
Yeah, I think "rewrite commit, but don't count it for the purpose of rebasing descendants/branches" might work. I'm still thinking about what the api could look like that would do all that's needed.